### ISSUE TITLE
[QUESTION] Evaluate f(x) only once in MOO example

### ISSUE THREAD

**Issue by martinlahr** [facebook/Ax#1034]
Hi,

in the MOO tutorial example using qNEHVI ([Link](https://ax.dev/tutorials/multiobjective_optimization.html)), the target function is evaluated TWICE to get the objectives (once for each objective). 
```
class MetricA(NoisyFunctionMetric):
    def f(self, x: np.ndarray) -> float:
        return float(branin_currin(torch.tensor(x))[0])
    
class MetricB(NoisyFunctionMetric):
    def f(self, x: np.ndarray) -> float:
        return float(branin_currin(torch.tensor(x))[1])
```
This is fine if the evaluation is inexpensive. However, in my case, the evaluation involves the training of a DNN which is rather expensive. How can I tell Ax both objective values with only one evaluation of f(x)?

Best regards,
Martin

**Comment by mpolson64** [comment]
Hi Martin. I'm happy to hear you're interested in using Ax for multi-objective optimization! What you're wanting to do here, ensuring your trials do not need to do extra work, and be done in a number of ways and should be straightforward.
Your first option would be to use the Service API as in the first four cells of the above tutorial. In this paradigm there is no need to fuss with `Metric` objects and we can simply train our DNN, get the metric values we care about, and pass them into `ax_client` like via `ax_client.complete_trial(trial_index=trial_index, raw_data={"a": 123, "b": 456})`. This method is easiest for most of our users and I suggest giving it a shot first.
The next option would be to use the `Metric` objections in conjunction with a `Runner`. You can read more about this setup here: https://ax.dev/tutorials/gpei_hartmann_developer.html . In this setup (which is typically used for online experiments -- that is experiments where we have some external system conducting trials) the `Runner` would be made to train your DNN in some online system and store the result, and the `Metric`s would each be made to fetch the results rather than conduct the expensive training job. This entire setup is very flexible and can allow for a high level of automation in big systems, but to be frank is quite a bit of work to set up. I suggest trying out `AxClient` and seeing if that meets your needs first.
Let me know if this was helpful or if you have any other questions :)

**Comment by Balandat** [comment]
The other option here would be to just cache the result of your function, sth like 

```
from functools import cache

@cache
def my_expensive_moo_eval_function(x):
   # do some expensive stuff (here is where your eval goes)
   return result  # this is a two-element array, or maybe a dict...

class MetricA(NoisyFunctionMetric):
    def f(self, x: np.ndarray) -> float:
        return float(my_expensive_moo_eval_function(x))[0])
    
class MetricB(NoisyFunctionMetric):
    def f(self, x: np.ndarray) -> float:
        return float(my_expensive_moo_eval_function(x))[1])
```

I don't think np arrays are hashable directly, so you may need to do some other wrapping (e.g. convert `tolist()` and use that to call your evaluation).

**Comment by martinlahr** [comment]
Oh, I thought I _had_ to use the developer API for qNEHVI. Thanks for clarifying that this is not the case. Am I right in assuming that defining multiple objectives defaults to MOO which defaults to qNEHVI? Also, is raw data in the form `{"a" : (a_val, None)}` equivalent to the `NoisyFunctionMetric` (i.e. the noise level is inferred)?

**Comment by saitcakmak** [comment]
> Am I right in assuming that defining multiple objectives defaults to MOO which defaults to qNEHVI? Also, is raw data in the form {"a" : (a_val, None)} equivalent to the NoisyFunctionMetric (i.e. the noise level is inferred)?

Yes & yes :)

**Comment by martinlahr** [comment]
OK, thanks. It's working now.
